src/transformers/models/bartpho/tokenization_bartpho.py

```diff
@@ -157,12 +157,22 @@ def __init__(
         self.sp_model.Load(str(vocab_file))
 
         # Load the reduced vocab
-        self.fairseq_tokens_to_ids = {"<s>": 0, "<pad>": 1, "</s>": 2, "<unk>": 3}
+
+        # Keep order of special tokens for backward compatibility
+        self.fairseq_tokens_to_ids = {
+            token: token_id
+            for token_id, token in enumerate(
+                dict.fromkeys(
+                    [str(bos_token), str(pad_token), str(eos_token), str(unk_token), str(sep_token), str(cls_token)]
+                ).keys()
+            )
+        }
```

Comment:
```suggestion
        self.fairseq_tokens_to_ids = {
            str(token): token_id
            for token_id, token in enumerate([bos_token, pad_token, eos_token, unk_token, sep_token, cls_token])
        }
```
I think this is equivalent, and less convoluted.

Reply:
This was indeed my first solution too, but I thought it was safer to propose a solution that would keep the same ids that were hardcoded before in the code.

Using the default values:
```python
bos_token="<s>"
eos_token="</s>"
sep_token="</s>"
cls_token="<s>"
unk_token="<unk>"
pad_token="<pad>"
mask_token="<mask>"

```
Here is what it would look like with your solution:
```python
dict_sylvain = {
    str(token): token_id
    for token_id, token in enumerate([bos_token, pad_token, eos_token, unk_token, sep_token, cls_token])
}
print(dict_sylvain)
# {'<s>': 5, '<pad>': 1, '</s>': 4, '<unk>': 3}
```
With mine:
```python
dict_lucile = {
    str(token): token_id
    for token_id, token in enumerate(
        dict.fromkeys([bos_token, pad_token, eos_token, unk_token, sep_token, cls_token]).keys()
    )
}
print(dict_lucile)
# {'<s>': 0, '<pad>': 1, '</s>': 2, '<unk>': 3} <- this is what was hardcoded before
```

Note that I've changed a little bit the format of this part of the code in [this commit](https://github.com/huggingface/transformers/pull/15328/commits/37a40df37903a932c2f951cbd33acb684246bae7) in order to only use once `str(Xxx)`

Reply:
I think it's worth it not to "miss" index values - it would also cause errors with the loop that follows this dictionary. 

To be honest, I'm not sure why the special tokens were hardcoded in 2 parts before (`'<s>',  '<pad>', '</s>', '<unk>'` on one side and `<mask>` on the other) and I don't think that the order matter for the implementation in the `transformers` library. I was just proposing something that would not changed the previous default behavior. If we don't care maybe the simpler is to do:
```python
dict_new = {
    str(token): token_id
    for token_id, token in enumerate(set([bos_token, pad_token, eos_token, unk_token, sep_token, cls_token, mask_token]))
}
print(dict_new)
# {'</s>': 0, '<unk>': 1, '<pad>': 2, '<s>': 3, '<mask>': 4}
```

What do you think?

Reply:
Ah, I understand the difference. So we can't get rid of the `str` here and the tokens have to arrive in this exact order, but we don't want to add duplicates. Maybe it's worth not doing a list comprehension then since this logic is hard to understand from the call to `dict.from_keys` and do a manual loop like:
```
tokens_to_id = {}
cnt = 0
for token in [bos_token, pad_token, eos_token, unk_token, sep_token, cls_token]:
    if str(token) not in tokens_to_id:
        tokens_to_id[str(token)] = cnt
        cnt += 1
```

Reply:
I absolutely understand your point! It will indeed be easier to understand this constraint with your proposal instead of using `dict.from_keys(..).keys()`! 

I have committed the changes [here](https://github.com/huggingface/transformers/pull/15328/commits/ec4ab8c195de7c77587301e6ce7d32b366f2a3b6). :smile:

Reply:
Thanks a lot!